cpp/tensorrt_llm/kernels/cutlass_kernels/include/moe_kernels.h

```diff
@@ -0,0 +1,846 @@
+        LoraImplPtr fc2_lora_impl, void* workspace, cudaEvent_t* memcpy_event_ptr,
+        int32_t const* gated_lora_ranks = nullptr, void const* const* gated_lora_weight_ptrs = nullptr)
+        : fc1_lora_ranks(fc1_lora_ranks)
+        , fc1_lora_weight_ptrs(fc1_lora_weight_ptrs)
+        , fc2_lora_ranks(fc2_lora_ranks)
+        , fc2_lora_weight_ptrs(fc2_lora_weight_ptrs)
+        , gated_lora_ranks(gated_lora_ranks)
+        , gated_lora_weight_ptrs(gated_lora_weight_ptrs)
+        , num_reqs(num_reqs)
+        , fc1_lora_impl(fc1_lora_impl)
+        , fc2_lora_impl(fc2_lora_impl)
+        , workspace(workspace)
+        , memcpy_event_ptr(memcpy_event_ptr)
+    {
+    }
+};
+
+namespace cutlass_kernels
+{
+static inline size_t pad_to_multiple_of_16(size_t const& input)
+{
+    static constexpr int ALIGNMENT = 16;
+    return ALIGNMENT * ((input + ALIGNMENT - 1) / ALIGNMENT);
+}
+
+class CubKeyValueSorter
+{
+public:
+    CubKeyValueSorter();
+
+    CubKeyValueSorter(int const num_experts_per_node);
+
+    void updateNumExperts(int const num_experts_per_node);
+
+    static size_t getWorkspaceSize(size_t const num_key_value_pairs, int const num_experts_per_node);
+
+    void run(void* workspace, size_t const workspace_size, int const* keys_in, int* keys_out, int const* values_in,
+        int* values_out, size_t const num_key_value_pairs, cudaStream_t stream);
+
+private:
+    static int expertsToBits(int experts);
+    int num_experts_;
+    int num_bits_;
+};
+
+/**
+ * \brief Describes what parallelism mode the MoE is using
+ *
+ * Tensor Parallelism refers to the mode where the weight matrices for each expert are sliced up between nodes.
+ * Each node will handle part of each expert, the final result is achieved by summing the result.
+ * The inter_size dimension should be divided by the number of nodes prior to passing it to the MoE plugin, only the
+ * required slice of the weights should be provided to the plugin FC1 is a ColumnLinear and FC2 is a RowLinear, see
+ * tensorrt_llm/mlp/mlp.py for an example of how this works for a single MLP
+ *
+ * NOTE: The bias for fc2 is only applied on rank 0. If we added it on all nodes the allreduce() would contain multiple
+ * copies of the bias. The bias on other node will be ignored, and may be set to nullptr
+ *
+ * Expert Parallelism refers to the mode where experts are divided between the nodes. Each node will handle only the
+ * tokens that are routed to the experts it is assigned to. Only the weights for the node's experts should be provided
+ * to the plugin For example, with #experts = 8, expert parallelism = 2: Node 0 would handle experts 0-3, and node 1
+ * would handle experts 4-7
+ *
+ * Regardless of parallelism mode:
+ *  * The input routing values must be the complete routing for all tokens/experts (required for softmax)
+ *  * An allreduce must be run on the result to combine the results from different nodes if parallelism > 1
+ */
+struct MOEParallelismConfig
```

Comment:
This should be outside the `cutlass_kernels` namespace and combined with internal so we dont need to do `using MOEParallelismConfig = ...`